common/http_test.go

```diff
@@ -0,0 +1,169 @@
+package common_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"sync"
+	"testing"
+
+	"github.com/tendermint/tmlibs/common"
+)
+
+func TestWriteSuccess(t *testing.T) {
+	w := httptest.NewRecorder()
+	common.WriteSuccess(w, "foo")
+	got, want := w.Code, 200
+	if got != want {
+		t.Errorf("got=%d; want=%d", got, want)
+	}
+}
+
+var blankErrResponse = new(common.ErrorResponse)
+
+func TestWriteError(t *testing.T) {
+	w := httptest.NewRecorder()
+	msg := "this is an error"
+	common.WriteError(w, errors.New(msg))
+	if g, w := w.Code, http.StatusBadRequest; g != w {
+		t.Errorf("got=%d; want=%d", g, w)
+	}
+	blob, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("ioutil.ReadAll err=%v", err)
+	}
+	recv := new(common.ErrorResponse)
+	if err := json.Unmarshal(blob, recv); err != nil {
+		t.Fatalf("json.Unmarshal err=%v", err)
+	}
+	if reflect.DeepEqual(recv, blankErrResponse) {
+		t.Errorf("expecting a non-blank error response")
+	}
+}
+
+func TestWriteCode(t *testing.T) {
+	codes := []int{
+		http.StatusOK,
+		http.StatusBadRequest,
+		http.StatusUnauthorized,
+		http.StatusInternalServerError,
+	}
+
+	for i, code := range codes {
+		w := httptest.NewRecorder()
+		common.WriteCode(w, "foo", code)
+		got, want := w.Code, code
+		if got != want {
+			t.Errorf("#%d: got=%d; want=%d", i, got, want)
+		}
+	}
+}
+
+type saver struct {
+	Foo int    `json:"foo" validate:"min=10"`
+	Bar string `json:"bar"`
+}
+
+type rcloser struct {
+	closeOnce sync.Once
+	body      *bytes.Buffer
+	closeChan chan bool
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func (rc *rcloser) Close() error {
+	var err = errAlreadyClosed
+	rc.closeOnce.Do(func() {
+		err = nil
+		rc.closeChan <- true
+		close(rc.closeChan)
+	})
+	return err
+}
+
+func (rc *rcloser) Read(b []byte) (int, error) {
+	return rc.body.Read(b)
+}
+
+var _ io.ReadCloser = (*rcloser)(nil)
+
+func makeReq(strBody string) (*http.Request, <-chan bool) {
+	closeChan := make(chan bool, 1)
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(strBody))
+	req := &http.Request{
+		Header: make(http.Header),
+		Body:   &rcloser{body: buf, closeChan: closeChan},
+	}
+	return req, closeChan
+}
+
+func TestParseRequestJSON(t *testing.T) {
+	tests := [...]struct {
```

Comment:
Good question. So using a slice is usually nice, but doesn't allow you to seek back to the index easily in case of any error, for example
given an error
```shell
sample_test.go:174 "#5: unknown string"
```

## Slice
```go
tests := []string{"a","b","c","d","e","f","g"}
```

## Indexed slice
```
tests := [...]string{0:"a", 1:"b", 2:"c", 3:"d", 4:"e", 5:"f", 6:"g"}
```
in the case of the slice, you'll have to manually go find the element then edit it, and in many cases since we are humans, might change the wrong element(testing becomes harder to fix)
in the case of indexed slice, you just search /5: and that's it. It's a personal style that I've adopted from dealing with many confusing tests cases.